### PR TITLE
Add randomization options to satin columns and rewrite split stitches

### PR DESCRIPTION
**Pull request by george-steel**
* Add a system for reproducible, randomized stitch behaviour using `hashlib.blake2s` as a pseudo-random oracle.
  * Unlike the stateful generator used in @claudinepeyrat06's patch, this prevents different randomized aspects from affecting each other, allowing editing while keeping most of the effects of the chosen seed.
* Introduce a `random_seed` param on EmbroideryElements and a button in the params window to re-roll the seed on each element individually.
* Add randomized options to satin columns, controlling both endpoints and split stitches.
  * expand and contract endpoints
  * zigzag spacing
  * Jitter for split stitch position
  * random-phase split stitches
* Make shortened stitches and split stitches compatible with each other, with centered split stitches based on before-shortening endpoints. Previously, split stitches disabled shortening.
* Prevent simulator window from eating exception messages from stitch plan computation. They now get correctly written to stderr.

### CODE REVIEW

**Comment by george-steel** [comment]
@tyrosinase I wanted to double-check with you since one of the changes here makes split satin stitches no linger inhibit short-stitch insetting. I made the 2 features play nice with each other (so if both were intended, they do not interfere and split stitches still make smooth divisions), but short stitch insetting is enabled by default, so this may produce undesired effects on designs that were relying on the inhibition. Does this break anything for you?

**Comment by george-steel** [comment]
@claudinepeyrat06 I haven't stitched it out yet, but I converted your bee to this version.

test-bee-converted

**Comment by tyrosinase** [comment]
> designs that were relying on the inhibition

Today I noticed that the option to turn it off existed so, I guess I haven't been using it. (I have noticed that the short stitches were generating once or twice, went "huh, that looks nice," and that's been about it. I usually structure my satins so they're not making the kind of shape that triggers it.)

Thanks for asking, though.

File: lib/elements/satin_column.py [hunk 1]
```diff
@@ -696,19 +770,23 @@ def walk(self, path, start_pos, start_index, distance):
...
-        def add_pair(pos0, pos1):
-            pos0, pos1 = self.offset_points(pos0, pos1, offset_px, offset_proportional)
-            points[0].append(pos0)
-            points[1].append(pos1)
+        # pre-cache ramdomised parameters to avoid property calls in loop
+        if use_random:
+            seed = prng.joinArgs(self.random_seed, "satin-points")
+            offset_proportional_min = np.array(offset_proportional) - self.random_width_decrease
+            offset_range = (self.random_width_increase + self.random_width_decrease)
+            spacing_range = spacing * self.random_zigzag_spacing / 100
```
Comment (kaalleen) [inline lib/elements/satin_column.py:784]: If I am not wrong you don't need to divide through 100 anymore (already done in line 120)

Comment (george-steel) [reply lib/elements/satin_column.py]: Good catch, fixed

**Comment by george-steel** [comment]
test bee
@claudinepeyrat06 I stitched out your bee (with some density adjustmenst to use the thread I had on hand) rendered on this version and it turned out quite nicely.

**Comment by claudinepeyrat06** [comment]
You do have added a tremendous amount of possibilities, thanks for that.
 I think that the random phase for split stitches  is going to be extremely useful for oversized fonts : i tried it with Lobster font at 400% (with also  a density decrease )
test 
it is much better that what i could do without this option. 

It's also  very nice to be able to have both split sapin and short stitch option (but it does change the way some of the lettering are embroidered, hopefully for the best).

I do have a question : Suppose  i create a random satin column and I do like the first result, so much that i want several copies  of it, can i do that ? if i duplicate, or copy and paste then the seed is different. If i reroll then i do see the value of the seed, and therefore i can have an exact copy, but otherwise the seed field is empty. More a question than a real concern.....

Also i'm wondering if it would be interesting to have random random parameters? sometimes more than rerolling with the same percentages, it would be great to change the percentage, for instance to create a bunch of similar but different flowers.

Your bee is great, i'm sorry i totally forgot  to send you mine, i was away from my files when i received the mail, and  then after i forgot...

**Comment by lexelby** [comment]
Awesome, great work!  I'm really excited about the possibilities this opens.

I'm curious about your hash-based PRNG.  What led you to choose hashing instead of a seeded PRNG?  And what are the properties of BLAKE2s that led you to choose it over other hash algorithms?

**Comment by george-steel** [comment]
> I'm curious about your hash-based PRNG. What led you to choose hashing instead of a seeded PRNG? And what are the properties of BLAKE2s that led you to choose it over other hash algorithms?

@lexelby I chose the pseudo-random oracle over the stateful generator as it can be used to generate multiple independent random streams by adding different prefixes to the counter. This allows the user to modify paths (which may change the number of times random numbers are chosen) and still keep the effects of the selected seed mostly the same, as different parts of the stitch algorithm do not effect each other's random rolls. Also, as the counters are explicit instead of existing in a thread-local, this approach makes the random output fairly immune to refactoring as the order of calls does not matter.

I chose blake2s as it is the fastest hash algorithm in python's standard library (on a 32-bit build such as our windows builds). It was also explicitly designed to be usable as a pseudo-random generator.

**Comment by lexelby** [comment]
Brilliant!  I'm so glad you're the one that implemented this, because I wouldn't have thought of using hashing to resolve those difficulties.

Could you add a comment in prng.py explaining some or all of this?  It wouldn't do to have someone try to "optimize" your code by removing the hashing.

**Comment by kaalleen** [comment]
I agree, this is so very nice. The next release is going to be great with so many new ceative possibilities.

**Comment by george-steel** [comment]
> Could you add a comment in prng.py explaining some or all of this? It wouldn't do to have someone try to "optimize" your code by removing the hashing.

Added. LGTY @lexelby?

File: lib/elements/satin_column.py [hunk 2]
```diff
@@ -84,13 +87,68 @@ def e_stitch(self):
...
+                     'Two values separated by a space may be used for an aysmmetric effect.'),
+           default=0, type='float', unit="% (each side)", sort_index=90)
+    @cache
+    def random_width_increase(self):
+        return self.get_split_float_param("random_width_increase_percent", (0, 0)) / 100
+
+    @property
+    @param('random_zigzag_spacing_percent',
+           _('Random zig-zag spacing percentage'),
+           tooltip=_('Afount of random jitter added to stitch length.'),
```
Comment (kaalleen) [inline lib/elements/satin_column.py:116]: Afount -> Amount

**Comment by george-steel** [comment]
@kaalleen @lexelby Is this ready to merge?

**Review by kaalleen — APPROVED** [review]
Looks good to me. Let's do it

**Comment by kaalleen** [comment]
I noticed, that since this branch has been merged to main that the center walk underlay behaves a little bit different from before.

center_walk

**Comment by lexelby** [comment]
Hmm, that's not good.  @george-steel what do you think?

**Comment by george-steel** [comment]
@kaalleen @lexelby Good catch. Found the bug and put up a fix in #2011.

**Pull request merged**